File: src/affix_helpers.ts
function mameCompatible(c: string): boolean {
    return vowel(c) || checkCharPresence(c, CONS_GROUP1);
}

function babeCompatible(c: string): boolean {
    return checkCharPresence(c, CONS_GROUP2);
}

function chooseMBP(c: string, softOffset: number, mAffixes: string[], bAffixes: string[], pAffixes: string[]): string {
    if (mameCompatible(c)) {
        return mAffixes[softOffset];
    }
    if (babeCompatible(c)) {
        return bAffixes[softOffset];
    }
    return pAffixes[softOffset];
}

function chooseLDT(c: string, softOffset: number, lAffixes: string[], dAffixes: string[], tAffixes: string[]): string {
    if (vowel(c) || checkCharPresence(c, CONS_GROUP7)) {
        return lAffixes[softOffset];
    }
    if (checkCharPresence(c, CONS_GROUP8)) {
        return dAffixes[softOffset];
    }
    return tAffixes[softOffset];
}

File: kotlin/grammar/src/main/kotlin/com/khairulin/kazakhverb/grammar/AffixHelpers.kt
package com.khairulin.kazakhverb.grammar

object AffixHelpers {
    fun mameCompatible(char: Char): Boolean {
        return Rules.VOWELS.contains(char) || Rules.CONS_GROUP1.contains(char)
    }

    fun babeCompatible(char: Char): Boolean {
        return Rules.CONS_GROUP2.contains(char)
    }

    fun chooseMBP(char: Char, softOffset: Int, mAffixes: List<String>, bAffixes: List<String>, pAffixes: List<String>): String {
        if (mameCompatible(char = char)) {
            return mAffixes[softOffset]
        }
        if (babeCompatible(char = char)) {
            return bAffixes[softOffset]
        }
        return pAffixes[softOffset]
    }

    fun chooseLDT(char: Char, softOffset: Int, lAffixes: List<String>, dAffixes: List<String>, tAffixes: List<String>): String {
        if (Phonetics.isVowel(char) || Rules.CONS_GROUP7.contains(char)) {
            return lAffixes[softOffset]
        }
        if (Rules.CONS_GROUP8.contains(char)) {
            return dAffixes[softOffset]
        }
        return tAffixes[softOffset]
    }
}